### ISSUE TITLE
Generate Kotlin code from static library

### ISSUE THREAD

**Issue by threema-theo** [mozilla/uniffi-rs#2291]
The docs read as follows:

> In our example, we generate the bindings with:
>
> ```
> cargo build --release
> cargo run --bin uniffi-bindgen generate --library target/release/libmath.so --language kotlin --out-dir out
> ```
>
>(maybe .dylib, good luck with .dll!)

More precisely: Running the above command on Linux works, while running it on Mac OS requires the library target to end with `.dylib`. 

As we're building code for an Android app, I don't see why the built library must be dynamic. 

To avoid differentiating the build platform, the cleanest way would therefore be to supply a static library (tell me if I'm wrong):

`Cargo.toml`:
```toml
...
[lib]
crate-type = ["lib", "cdylib", "staticlib"]
...
```

We could then build with
 ```
cargo build --release
cargo run --bin uniffi-bindgen generate --library target/release/libmath.a --language kotlin --out-dir out
 ```

However, this does not work due the following code:
https://github.com/mozilla/uniffi-rs/blob/09bc9e9084dc4d12cad2816df84f5ce2ed013e6f/uniffi/src/cli/uniffi_bindgen.rs#L167-L173

Why is this check needed? Could we remove this lines (or make an exception for kotlin)? 
I tested the kotlin generation without this check and did not face any issues yet.

Thanks for your insights!

**Comment by mhammond** [comment]
The foreign languages need to load the .so at runtime to locate the public functions. Even Swift works like this.

The code you referenced isn't actually used at runtime, but at build time to locate all the metadata emitted by uniffi - it is necessary for swift to be able to load these from the partially built .a files, but it still needs a dylib at runtime. While this might be able to made work for Kotlin, it's not typically how Kotlin is built given you must end up with a .so.

To put it another way, that check there is mainly to *prevent* people trying to build static libs with Kotlin because it's not going to end up working - you will be able to get past this, but no further with a static lib.

**Comment by threema-theo** [comment]
Thanks a lot for your answer.

So the way to streamline the build process for an Android app for different platforms might be to use [cargo-ndk](https://github.com/bbqsrc/cargo-ndk).
